File: src/main/java/gargoyle/xenox/util/log/StreamLogger.java
package gargoyle.xenox.util.log;

import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.Writer;
import java.nio.charset.Charset;

public abstract class StreamLogger extends Logger {
    private final PrintWriter writer;

    StreamLogger() {
        this(Charset.defaultCharset());
    }

    private StreamLogger(Charset charset) {
        this(new OutputStreamWriter(System.err, charset));
    }

    private StreamLogger(Writer writer) {
        super(null);
        this.writer = new PrintWriter(writer);
    }

    @Override
    protected final synchronized void log(StackTraceElement caller, long dateTime, LEVEL level, String message, Throwable throwable) {
        String msg = String.format("[%1$tF %1$tC] [%2$s] %3$s: %4$s%n%5$s", dateTime, level,
                String.format("%s.%s(%s:%d)", caller.getClassName(), caller.getMethodName(), caller.getFileName(), caller.getLineNumber()),
                message, getStackTraceAsString(throwable)).trim();
        writer.println(msg);
        writer.flush();
    }
}

File: src/main/kotlin/gargoyle/xenox/util/log/StreamLogger.kt
package gargoyle.xenox.util.log

import java.io.OutputStreamWriter
import java.io.PrintWriter
import java.io.Writer
import java.nio.charset.Charset

abstract class StreamLogger private constructor(writer: Writer) : Logger(LEVEL.DEBUG) {
    private val writer: PrintWriter

    internal constructor() : this(Charset.defaultCharset())
    private constructor(charset: Charset) : this(OutputStreamWriter(System.err, charset))

    init {
        this.writer = PrintWriter(writer)
    }

    override fun log(
        caller: StackTraceElement,
        dateTime: Long,
        level: LEVEL,
        message: String,
        throwable: Throwable?
    ) {
        val msg = String.format(
            "[%1\$tF %1\$tC] [%2\$s] %3\$s: %4\$s%n%5\$s",
            dateTime,
            level,
            String.format("%s.%s(%s:%d)", caller.className, caller.methodName, caller.fileName, caller.lineNumber),
            message,
            getStackTraceAsString(throwable)
        ).trim()
        writer.println(msg)
        writer.flush()
    }
}
